Revisions post meetings routing comp edits (#554)
opendcs/opendcs: docs/source/start-routingspec.rst

files.

The routing spec guide and steps below assume that a user already has an LRGS
connection set up and is operational.  In other words, that means that messages are 
being retrieved and are available for routing specs or using the DCP message 
browser.  For more information about getting started with the LRGS see the 
legacy LRGS guide :doc:`LRGS Guide <./legacy-lrgs-userguide>`.

A couple checks to see if the LRGS is running are outlined below.

.. table:: Table of LRGS Checks

   +-------------------------------+----------------------------------+-----------------------+
   | **Check**                     | **Typical Location or Command**  | **What to Look For**  |
   +===============================+==================================+=======================+
   | Check **lock** file           | <path>/opendcs_user/lrgs/        | Existence             |
   +-------------------------------+----------------------------------+-----------------------+
   | Check **archive** folder      | <path>/opendcs_user/lrgs/archive/| Recently Files        |
   +-------------------------------+----------------------------------+-----------------------+

> Additionally, Unix commands can be run the Unix serves to help to identify whether 
> the continuous LRGS process is running:
> 

Reviewer: ```suggestion
The following Unix command can be run to determine if the LRGS process is running:

```

* **web** - reads data files over a web connection specified by a URL
* **abstractweb** - reads data files over a web connection specified by a URL with parameters
* **socketstream** - opens a socket and reads a one-way stream of data containing raw DCP messages. Some DRGS and HRIT product provide such a stream

> * **hotbackupgroup** - an ordered group of LRGS data sources, where secondary and tertiary servers are used with primary or secondary is unavailable, respectively
> * **roundrobingroup** - contains a list of other data sources and is continually read in
> 

Reviewer: ```suggestion
* **hotbackupgroup** - an ordered group of LRGS data sources, where secondary and tertiary servers are used to automatically maintain availability
* **roundrobingroup** - contains a list of other data sources and is continually read in

```

* Sample Message: Box for pasting or loading retrieved messages to test DECODING
* Decoded Data: Output of DECODING with color syntax

In addition to the parts highlighted above, there are three other parts that 
the user may use:

* Sensor Line Conversions: Table for specifying units of decoded message and any simple linear coefficient conversions
* Data order:  A drop down menu where Ascending or Descending can be selected (default is undefined)
* Header Type:  A drop down menu for selecting a header type such as a medium or source type (default is blank)

In regards to using the sample message browser, note that messages
can only be retrieved from an lrgs for which a connection has already
been established.  For USACE users, this means that messages can only

> be retrieved while logged onto the server.

Reviewer: Consider dropping this sentence about USACE users.   It is not clear to me what the 'server' means in this context.

Author: I think this will be clear to USACE users - but perhaps this something that could be elaborated on and explained in the Confluence Page - given the relevance to USACE/CWMS?

+---------------------------+---------------------+--------------------------------------------------------+

DECODING EXAMPLES
~~~~~~~~~~~~~~~~~

Below is an example of when the message is formatted like a csv.  

> The second examples shows a simple GOES dcp message with one variable.

Reviewer: ```suggestion
The second examples shows a simple GOES DCP message with one variable.
```

Once the new window comes up the user should address the following fields, at least.

+-------------------------+------------------------------------------------+
|**Field**                |**Description**                                 |
+=========================+================================================+
|Data Source              |Location in Database                            |
+-------------------------+------------------------------------------------+
|Destination              |consumer (pipe, database, file, etc)            |
+-------------------------+------------------------------------------------+
|Output Format            |set output format                               |
+-------------------------+------------------------------------------------+
|Date/Time - Since/Util   |define time window for message retrieval        |
+-------------------------+------------------------------------------------+
|Platform Selection       |Specify a platform or network list              |
+-------------------------+------------------------------------------------+

> 

Reviewer: The first sentence is rendering as a Title.
![image](https://github.com/opendcs/opendcs/assets/4818531/7d45086d-26dc-49be-a75f-ed3407bdedd7)

Author: Right now it's looking reasonable format on my end - let me know if this is still showing up goofy? 
![image](https://github.com/opendcs/opendcs/assets/104451600/33f260c3-33ef-432d-ab2b-f3db6b5e624a)

The **Data Source** will populate with a drop down menu consisting of the

>  source or sources added in step 3.  The **Destination** will populate with 

Reviewer: ```suggestion
sources previously added in the 'Sources' tab.  The **Destination** will populate with 
```

 some options including but not limited to the database, file and pipe.  
 To set up a routing spec to go into the database, select the applicable 
 database.  The **Date/Time** format is only applicable for sources

>  that are not static.  For example, sources of type lrgs or abstract web

Reviewer: suggestion: 
 The **Date/Time** selection is used with Data Sources
 that require selecting a range of Date/Time.

Date/Time is ignored by 'static' Data Sources, that provide a pre-defined time window; such as always providing the last twenty four hours of data.

 will require a look back time window to be defined.  The longer the look 
 back the longer the routing spec will take to run, because there will
 presumably be more data.  It is good practice to take into consideration
 how frequently the source is updated, the resolution of the data, 
 how frequently you are planning to run the routing spec, and how many
 platforms the routing spec will include.  The **Platform Selection** 
 is where the user will need to either specify a platform or network list.
 Recall that a network list is a group of platforms.  

.. image:: ./media/start/routingspec/im-28-rs-source.JPG
   :alt: new list window
   :width: 500

NOTE - there are some unique criteria that need to be selected given 
certain types of sources.  For example, when the source type is a 

> directory, then the property "oneScanOnly" must be defined as True.

Reviewer: ```suggestion
directory, then the property "oneScanOnly" can be defined as True 
to prevent continuously scanning the directory.
```

Author: At least in the version LRE has running - this HAS to be set as True.  Maybe perhaps better said if it is not set as True, the routing spec will NEVER close and there will be never ending scanning of the directory.

TIP - It is good practice to define the **Destination** as pipe before
putting a routing spec in operation.  This allows the user to test the 
routing spec without potentially overwriting or accidentally populating
another time series in the database.  

Remember to save the changes.  Now you have set up a routing spec.  
This means that assuming all the steps above have been completed
or checked off, then a user can now run the routing spec. Routing
specs can be run manually, irregardless of whether any continuous 
OpenDCS process is running.  If the destination is a database, the
database must be up and running.  

How do I run or execute the Routing Spec?
-----------------------------------------

> Thus far, we have used the OpenDCS GUI to set up the routing spec.  
> However, running the routing spec 'manually' requires running the
> OpenDCS command 'rs'.  In the following examples, imagine there is 
> a routing spec named 'mySpec'.
> 

Reviewer: ```suggestion
Thus far, we have used the OpenDCS GUI to set up the routing spec.  
However, running the routing spec 'manually' requires running the
OpenDCS command 'rs'.  In the table below each option for the rs command is demonstrated with a routing spec named 'mySpec'.   In practice multiple options 
are often used to achieve the desired result such as using both the 'Since Time' and 'Until Time' options.

```

+------------------------------+-----------------------------------------------------------------+

Manual Routing Specs Common Options
-----------------------------------

Troubleshooting - Use Debugger
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

... more content coming soon ...

Backfilling - Use Since and Until
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

... more content coming soon ...

Lockfile Significance
~~~~~~~~~~~~~~~~~~~~~

... more content coming soon ...

9.Routing Scheduler
===================

Now that you have set up a routing spec, you may wish to automate it.  
This is so that a user does not have to manually run the routing spec
as described above.  Depending on the system where the database is running,
and where the continuous processes are going to run, the following methods 
can be used to automate the scheduling of the routing specs.  The first
method is using an appropriate system job scheduler. This is independent 
of the OpenDCS software. A few options are outlined below.

System Scheduler
----------------

For a Windows System the following job schedulers are options:
* Windows Scheduler

> * Python Crond

Reviewer: Python is not built into windows.  
do you mean: https://pypi.org/project/python-crontab/ ?
